File: Projects/Projects/src/co/edu/uptc/java/doubleList/DoubleList.java
package co.edu.uptc.java.doubleList;

public class DoubleList<T> {

    private NodeDoubleList<T> head;
    private int size;

    public DoubleList() {
        head = null;
        size = 0;
    }

    public int size() {

        if (size == Integer.MAX_VALUE) {
            return Integer.MAX_VALUE;
        }

        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean contains(Object o) {

        if (o == null) {
            throw new NullPointerException("La lista no permite datos nulos");
        }

        NodeDoubleList<T> auxNode = head;

        while (auxNode != null) {

            if (auxNode.getValue().equals(o)) {
                return true;
            }

            auxNode = auxNode.getNext();
        }

        return false;
    }

    public Object[] toArray() {

        if (isEmpty()) {
            return new Object[0];
        }

        Object[] array = new Object[size];

        NodeDoubleList<T> current = head;

        int index = 0;

        while (current != null) {

            array[index] = current.getValue();

            current = current.getNext();

            index++;
        }

        return array;
    }

    public boolean add(T e) {

        NodeDoubleList<T> newNode = new NodeDoubleList<>(e);

        if (head == null) {

            head = newNode;
        }
        else {

            NodeDoubleList<T> current = head;

            while (current.getNext() != null) {
                current = current.getNext();
            }

            current.setNext(newNode);

            newNode.setPrevious(current);
        }

        size++;

        return true;
    }

    public boolean remove(Object o) {

        if (head == null) {
            return false;
        }

        if (head.getValue().equals(o)) {

            head = head.getNext();

            if (head != null) {
                head.setPrevious(null);
            }

            size--;

            return true;
        }

        NodeDoubleList<T> aux = head;

        while (aux != null) {

            if (aux.getValue().equals(o)) {

                aux.getPrevious().setNext(aux.getNext());

                if (aux.getNext() != null) {
                    aux.getNext().setPrevious(aux.getPrevious());
                }

                size--;

                return true;
            }

            aux = aux.getNext();
        }

        return false;
    }

    public T get(int index) {

        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException("Índice fuera de rango: " + index);
        }

        NodeDoubleList<T> aux = head;

        for (int i = 0; i < index; i++) {
            aux = aux.getNext();
        }

        return aux.getValue();
    }
}

File: BenchmarkCsharp/dobleList/DoubleList.cs
using System;

namespace Co.Edu.Uptc.Csharp.DoubleList
{
    public class DoubleList<T>
    {
        private NodeDoubleList<T>? head;
        private int size;

        public DoubleList()
        {
            head = null;
            size = 0;
        }

        public int Size()
        {
            if (size == int.MaxValue)
            {
                return int.MaxValue;
            }
            return size;
        }

        public bool IsEmpty()
        {
            return size == 0;
        }

        public bool Contains(object? o)
        {
            if (o == null)
            {
                throw new ArgumentNullException(nameof(o), "La lista no permite datos nulos");
            }

            NodeDoubleList<T>? auxNode = head;

            while (auxNode != null)
            {
                if (auxNode.Value!.Equals(o))
                {
                    return true;
                }

                auxNode = auxNode.Next;
            }

            return false;
        }

        public object?[] ToArray()
        {
            if (IsEmpty())
            {
                return Array.Empty<object>();
            }

            object?[] array = new object?[size];

            NodeDoubleList<T>? current = head;

            int index = 0;

            while (current != null)
            {
                array[index] = current.Value;
                current = current.Next;
                index++;
            }

            return array;
        }

        public bool Add(T e)
        {
            NodeDoubleList<T> newNode = new NodeDoubleList<T>(e);

            if (head == null)
            {
                head = newNode;
            }
            else
            {
                NodeDoubleList<T>? current = head;

                while (current!.Next != null)
                {
                    current = current.Next;
                }

                current.Next = newNode;
                newNode.Previous = current;
            }

            size++;

            return true;
        }

        public bool Remove(object? o)
        {
            if (head == null)
            {
                return false;
            }

            if (head.Value!.Equals(o))
            {
                head = head.Next;

                if (head != null)
                {
                    head.Previous = null;
                }

                size--;

                return true;
            }

            NodeDoubleList<T>? aux = head;

            while (aux != null)
            {
                if (aux.Value!.Equals(o))
                {
                    aux.Previous!.Next = aux.Next;

                    if (aux.Next != null)
                    {
                        aux.Next.Previous = aux.Previous;
                    }

                    size--;

                    return true;
                }

                aux = aux.Next;
            }

            return false;
        }

        public T Get(int index)
        {
            if (index < 0 || index >= size)
            {
                throw new ArgumentOutOfRangeException(nameof(index), $"Índice fuera de rango: {index}");
            }

            NodeDoubleList<T>? aux = head;

            for (int i = 0; i < index; i++)
            {
                aux = aux!.Next;
            }

            return aux!.Value;
        }
    }
}